dashboard/app/models/contact_rollups_pardot_memory.rb

```diff
@@ -27,12 +27,90 @@ class ContactRollupsPardotMemory < ApplicationRecord
   def self.add_and_update_pardot_ids(last_id = nil)
     last_id ||= ContactRollupsPardotMemory.maximum(:pardot_id) || 0
 
-    # TODO: bulk insert and update. or even bulk delete and then bulk insert
+    # TODO: save records in batch using activerecord_import upsert!
     PardotV2.retrieve_new_ids(last_id).each do |mapping|
       pardot_record = find_or_initialize_by(email: mapping[:email])
       pardot_record.pardot_id = mapping[:pardot_id]
       pardot_record.pardot_id_updated_at = Time.now
       pardot_record.save
     end
   end
+
+  def self.create_new_pardot_contacts
+    # Finds contacts in contact_rollups_processed but are not in contact_rollups_pardot_memory.
+    # Contacts must not be previously rejected by Pardot as invalid emails.
+    new_contacts_query = <<-SQL.squish
+      SELECT processed.email, processed.data
+      FROM contact_rollups_processed AS processed
+      LEFT OUTER JOIN contact_rollups_pardot_memory AS pardot
+        ON processed.email = pardot.email
+      WHERE pardot.pardot_id IS NULL
+        AND NOT (pardot.data_rejected_reason <=> '#{PardotHelpers::ERROR_INVALID_EMAIL}')
+    SQL
+
+    # Adds contacts to a batch and then sends batch requests to create new Pardot prospects.
+    # Requests may not be sent immediately until batch size is big enough.
+    ActiveRecord::Base.connection.exec_query(new_contacts_query).each do |record|
+      data_col = JSON.parse(record['data']).deep_symbolize_keys
+      contact = {email: record['email'], pardot_id: record['pardot_id']}.merge(data_col)
+
+      submissions, errors = PardotV2.batch_create_prospects contact
+      save_sync_results(submissions, errors, Time.now) if submissions.present?
+    end
+
+    # Sends Pardot request for the rest of the batch
+    submissions, errors = PardotV2.batch_create_remaining_prospects
```

Comment:
curious: what are the remaining prospects/why don't they get created by the first method?

Reply:
makes sense, thanks!